src/twisted/internet/test/test_inlinecb.py

```diff
@@ -3,22 +3,332 @@
 # See LICENSE for details.
 
 """
-Tests for L{twisted.internet.defer.inlineCallbacks}.
-
-Some tests for inlineCallbacks are defined in L{twisted.test.test_defgen} as
-well.
+Tests for L{twisted.internet.inlineCallbacks}.
 """
 
+import traceback
+import unittest as pyunit
+import weakref
+from typing import Any, Generator, List, Set, Union
 
+from twisted.internet import reactor, task
 from twisted.internet.defer import (
     CancelledError,
     Deferred,
+    Failure,
+    fail,
     inlineCallbacks,
     returnValue,
+    succeed,
 )
+from twisted.python.compat import _PYPY
 from twisted.trial.unittest import SynchronousTestCase, TestCase
 
 
+def getThing():
+    d = Deferred()
+    reactor.callLater(0, d.callback, "hi")
+    return d
+
+
+def getDivisionFailure(msg: Union[str, None] = None) -> Failure:
+    """
+    Make a L{Failure} of a divide-by-zero error.
+    """
+    try:
+        raise ZeroDivisionError(msg)
+    except BaseException:
+        f = Failure()
+    return f
+
+
+class TerminalException(Exception):
+    pass
+
+
+class BasicTests(TestCase):
+    # First provide all the generator impls necessary for BaseDefgenTests
+
+    @inlineCallbacks
+    def _genBasics(self):
+        x = yield getThing()
+
+        self.assertEqual(x, "hi")
+
+        try:
+            yield getDivisionFailure("OMG")
+        except ZeroDivisionError as e:
+            self.assertEqual(str(e), "OMG")
+        returnValue("WOOSH")
+
+    def testBasics(self):
+        """
+        Test that a normal inlineCallbacks works.  Tests yielding a
+        deferred which callbacks, as well as a deferred errbacks. Also
+        ensures returning a final value works.
+        """
+
+        return self._genBasics().addCallback(self.assertEqual, "WOOSH")
+
+    @inlineCallbacks
+    def _genBuggy(self):
+        yield getThing()
+        1 / 0
+
+    def testBuggy(self):
+        """
+        Ensure that a buggy generator properly signals a Failure
```

Comment:
What is a "buggy generator" ? What "properly" means?

Can this docstring be more specific ?

----------

What is my understanding  of the purpose of the tests, by reading the test code


"""
Any exception raised by a generator is converted into a L{Failure}.
"""

Is my understanding correct ?

Reply:
Correct. I've expanded the description, renamed the test and renamed _genBuggy in both inlinecb and defgen test suites.